org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLCodeActions.java

```diff
@@ -39,37 +42,56 @@ public class XMLCodeActions {
 	private static final Logger LOGGER = Logger.getLogger(XMLCompletions.class.getName());
 
 	private final XMLExtensionsRegistry extensionsRegistry;
+	
+	public static String NO_ERROR = "NoError";
 
 	public XMLCodeActions(XMLExtensionsRegistry extensionsRegistry) {
 		this.extensionsRegistry = extensionsRegistry;
 	}
 
 	public List<CodeAction> doCodeActions(CodeActionContext context, Range range, DOMDocument document,
 			SharedSettings sharedSettings, CancelChecker cancelChecker) {
-		cancelChecker.checkCanceled();
+		try {
+			cancelChecker.checkCanceled();
+	
+			List<CodeAction> codeActions = new ArrayList<>();
+			List<Diagnostic> diagnostics = context.getDiagnostics();
 
-		List<CodeAction> codeActions = new ArrayList<>();
-		List<Diagnostic> diagnostics = context.getDiagnostics();
-		if (diagnostics != null) {
-			for (Diagnostic diagnostic : context.getDiagnostics()) {
+			// Add a "No Error" diagnostic for two reasons:
+			// 1) In order to allow providing Code Actions for document offsets where no Error/sWarnings/Info 
+			//    diagnostic code exist
+			// 2) In order to prevent duplication of Code Actions that do not require a certain diagnostic code.
+			//    All code action participants that do not require a certain diagnostics node should react only to 
+			//    "NoError" diagnostic code - thus preventing the creation of duplicated code actions..
+			diagnostics = diagnostics != null ? diagnostics.stream().collect(Collectors.toList()) : new ArrayList<>();
+			Diagnostic noErrorDiagnostic = new Diagnostic(range, "No Error", DiagnosticSeverity.Hint, null, NO_ERROR);
+			if(!diagnostics.contains(noErrorDiagnostic)) {
+				diagnostics.add(noErrorDiagnostic);
+			}
+
+			for (Diagnostic diagnostic : diagnostics) {
 				for (ICodeActionParticipant codeActionParticipant : extensionsRegistry.getCodeActionsParticipants()) {
+					cancelChecker.checkCanceled();
 					try {
-						cancelChecker.checkCanceled();
 						CodeActionRequest request = new CodeActionRequest(diagnostic, range, document,
 								extensionsRegistry, sharedSettings);
 						codeActionParticipant.doCodeAction(request, codeActions, cancelChecker);
 					} catch (CancellationException e) {
-						throw e;
+						LOGGER.log(Level.FINER, e.getMessage(), e);
+						return Collections.emptyList();

```

Comment:
Why you dont rethrrow the cancel exception? All another participant like hover completion do that.

I did that to notify to the client that opération has been canceled

Reply:
@angelozerr Thanks for asking this. I also wanted to discuss this...

My short answer is: "because:
1) Participants API methods no not declare in their signatures that any exception like `CancellationException` can be thrown. 
2) We don't need to report anything when cancelled (imho - see my thoughts below...)"

In LemMinX-Maven hovers and definitions I don't re-throw any `CanceletionExceptions` -s - any "cancel" makes it to return a 'null" or empty result as soon as possible. I haven't looked into completion or other participants yet.

I don't think the cancellation is to be reported back to the client (especially in case if it was initiated by a client).

IMHO, if a client sends a cancellation request to the server - the server should close its current operation as soon as possible and not initiate sending of any extra statuses back to the client (and client, once it cancelled the current operation, is not expecting any valuable operation result nor any "prove" of that operation is really cancelled). If a server needs to cancel an ongoing operation - it's the sane - return "null" or empty result to a  client as soon as possible and that's enough.

Still not sure if it's really related, but we were already fixing the issue of such "cancellation reporting" - https://github.com/eclipse/lsp4e/pull/603

Also, leaving the `throw (CancellationException)e` here makes us to re-develop the tests so they correctly catch and process these exceptions - it's not the current case right now, so I stopped creating a JUnit test and make changes for  Cancellation Support in LemMinX-Maven diagnostic participant for the same reason: `o.e.lemminx.services.XMLDiagnostics.doDiagnostics(DOMDocument, XMLValidationSettings, Map<String, Object>, CancelChecker)`  throws `CancellationException` whithout any declaration of that this exception can be thrown. 

If I'm wrong and we still really need to throw and re-throw the `CancellationException` from Diagnostic/Definition/Hover/Content assist/Code action/etc participants, then I insist in adding  `throws CancellationException`   to their public API method signatures so everybody knows that this exception need to be handled.

Reply:
@vrubezhny if you see the doc of LSP4J you can see that the call of cancelChecker which can throw a CancellableException is not catched https://github.com/eclipse-lsp4j/lsp4j/blob/79df5e2ac119e0238a95567bb4bc0f1741428343/documentation/jsonrpc.md?plain=1#L67

Reply:
> If I'm wrong and we still really need to throw and re-throw the CancellationException from Diagnostic/Definition/Hover/Content assist/Code action/etc participants, then I insist in adding throws CancellationException to their public API method signatures so everybody knows that this exception need to be handled.

I think it is better to throw a CancellationException instead returning an empty list, but I suggest that you create an issue to LSP4J to clarify that. Once it is clarified, we will need to update APIs (define CancellationException  in the API as suggested or return an empty value for other operations).